include/seqan3/search/algorithm/search_result_range.hpp

```diff
@@ -0,0 +1,324 @@
+    {
+        return *current_buffer_it;
+    }
+
+    //!\brief Returns a pointer to the current search result.
+    pointer operator->() const noexcept
+    {
+        return std::addressof(*current_buffer_it);
+    }
+    //!\}
+
+    /*!\name Increment operators
+     * \{
+     */
+    //!\brief Increments the iterator by one.
+    search_result_range_iterator & operator++(/*pre*/)
+    {
+        assert(range_ptr != nullptr);
+
+        if (++current_buffer_it == last_buffer_it)
+            fetch_next_query_results();
+
+        return *this;
+    }
+
+    //!\brief Increments the iterator by one.
+    void operator++(int /*post*/)
+    {
+        ++(*this);
+    }
+    //!\}
+
+    /*!\name Comparison operators
+     * \{
+     */
+    //!\brief Checks whether `lhs` is equal to the sentinel.
+    friend constexpr bool operator==(search_result_range_iterator const & lhs,
+                                     std::ranges::default_sentinel_t const &) noexcept
+    {
+        return lhs.at_end && (lhs.current_buffer_it == lhs.last_buffer_it);
+    }
+
+    //!\brief Checks whether `lhs` is equal to `rhs`.
+    friend constexpr bool operator==(std::ranges::default_sentinel_t const & lhs,
+                                     search_result_range_iterator const & rhs) noexcept
+    {
+        return rhs == lhs;
+    }
+
+    //!\brief Checks whether `lhs` is not equal to the sentinel.
+    friend constexpr bool operator!=(search_result_range_iterator const & lhs,
+                                     std::ranges::default_sentinel_t const & rhs) noexcept
+    {
+        return !(lhs == rhs);
+    }
+
+    //!\brief Checks whether `lhs` is not equal to `rhs`.
+    friend constexpr bool operator!=(std::ranges::default_sentinel_t const & lhs,
+                                     search_result_range_iterator const & rhs) noexcept
+    {
+        return !(lhs == rhs);
+    }
+    //!\}
+
+private:
+    /*!\brief Fetches the next query results and updates the corresponding iterators to the new search result buffer.
+     *
+     * \details
+     *
+     * If the last search did not generate any hits, the next query is searched. This process is repeated
+     * until either a query produced at least one hit or all queries have been searched.
+     */
+    void fetch_next_query_results()
+    {
+        while (!(range_ptr->next() || range_ptr->at_end()));
+
+        current_buffer_it = std::ranges::begin(range_ptr->result_buffer);
+        last_buffer_it = std::ranges::end(range_ptr->result_buffer);
+
+        at_end = range_ptr->at_end();
+    }
+
+    //!\brief The current iterator over the buffered search results.
+    result_buffer_iterator current_buffer_it{};
+    //!\brief The end iterator over the buffered search results.
+    result_buffer_iterator last_buffer_it{};
+    //!\brief Pointer to the underlying range.
+    search_result_range * range_ptr{};
+    //!\brief Indicates the end of the underlying range.
+    bool at_end{true};
```

Comment:
No easily, since the comparison operator is a friend and has no access to the range the iterator is pointing to. Making the operator a friend of the range class does not seem worth the change. Besides, this is an input iterator, which will be move-only anyway. So size does not really matter.